api/api_errors.go

```diff
@@ -110,3 +115,22 @@ func (errF3ParticipationTicketStartBeforeExisting) Error() string {
 type errF3NotReady struct{}
 
 func (errF3NotReady) Error() string { return "f3 isn't yet ready to participate" }
+
+type ErrExecutionRevertedWithData struct {
+	message string
+	data    string
+}
+
+// Error returns the error message.
+func (e *ErrExecutionRevertedWithData) Error() string { return e.message }
+
+// ErrorData returns the error data.
+func (e *ErrExecutionRevertedWithData) ErrorData() interface{} { return e.data }
+
+// NewErrExecutionRevertedWithData creates a new ErrExecutionRevertedWithData.
+func NewErrExecutionRevertedWithData(data string) *ErrExecutionRevertedWithData {
+	return &ErrExecutionRevertedWithData{
```

Comment:
```suggestion
func (e ErrExecutionRevertedWithData) Error() string { return e.message }

// ErrorData returns the error data.
func (e ErrExecutionRevertedWithData) ErrorData() interface{} { return e.data }

// NewErrExecutionRevertedWithData creates a new ErrExecutionRevertedWithData.
func NewErrExecutionRevertedWithData(data string) ErrExecutionRevertedWithData {
	return ErrExecutionRevertedWithData{
```

Does jsonrpc bork still on non-interface? Not a big deal but if we can avoid pointer receivers where they are not necessary.

Reply:
@rvagg They are necessary as of now because we are registering the errors as `RPCErrors.Register(EExecutionRevertedWithData, new(*ErrExecutionRevertedWithData))
}` so `go-jsonrpc` will expect error to be returned as a pointer type for  `c, ok := e.byType[reflect.TypeOf(err)]; `.

I am trying to figure out a way, so we can get the code without worrying if the error is pointer type or not.

Reply:
no, don't bother, pointer is OK if you're forced to downstream